src/wasi/WASI.cpp

```diff
@@ -120,77 +175,73 @@ void WASI::fd_seek(ExecutionState& state, Value* argv, Value* result, Instance*
+    uint32_t length = argv[3].asI32();
     uint32_t oflags = argv[4].asI32();
     uint64_t rights = argv[5].asI64();
     uint64_t right_inheriting = argv[6].asI64();
     uint32_t fdflags = argv[7].asI32();
-    uint32_t ret_fd_offset = argv[8].asI32();
-
-    uvwasi_fd_t* ret_fd = reinterpret_cast<uvwasi_fd_t*>(instance->memory(0)->buffer() + ret_fd_offset);
-
-    const char* path = reinterpret_cast<char*>(instance->memory(0)->buffer() + path_offset);
-
-    result[0] = Value(static_cast<uint16_t>(
-        uvwasi_path_open(WASI::g_uvwasi,
-                         fd,
-                         dirflags,
-                         path,
-                         len,
-                         oflags,
-                         rights,
-                         right_inheriting,
-                         fdflags,
-                         ret_fd)));
+    const char* path = reinterpret_cast<char*>(get_memory_pointer(instance, argv[2], length));
+    uvwasi_fd_t* ret_fd = reinterpret_cast<uvwasi_fd_t*>(get_memory_pointer(instance, argv[8], sizeof(uvwasi_fd_t)));
+
+    result[0] = Value(uvwasi_path_open(WASI::g_uvwasi, fd, dirflags, path, length,
+                                       oflags, rights, right_inheriting, fdflags, ret_fd));
 }
 
 void WASI::environ_sizes_get(ExecutionState& state, Value* argv, Value* result, Instance* instance)
 {
-    uint32_t count = argv[0].asI32();
-    uint32_t buf = argv[1].asI32();
-
-    uvwasi_size_t* uvCount = reinterpret_cast<uvwasi_size_t*>(instance->memory(0)->buffer() + count);
-    uvwasi_size_t* uvBufSize = reinterpret_cast<uvwasi_size_t*>(instance->memory(0)->buffer() + buf);
+    uvwasi_size_t* uvCount = reinterpret_cast<uvwasi_size_t*>(get_memory_pointer(instance, argv[0], sizeof(uint32_t)));
+    uvwasi_size_t* uvBufSize = reinterpret_cast<uvwasi_size_t*>(get_memory_pointer(instance, argv[1], sizeof(uint32_t)));
 
-    result[0] = Value(static_cast<uint16_t>(uvwasi_environ_sizes_get(WASI::g_uvwasi, uvCount, uvBufSize)));
+    result[0] = Value(uvwasi_environ_sizes_get(WASI::g_uvwasi, uvCount, uvBufSize));
 }
 
 void WASI::environ_get(ExecutionState& state, Value* argv, Value* result, Instance* instance)
 {
-    uint32_t env = argv[0].asI32();
-    uint32_t environBuf = argv[1].asI32();
+    uvwasi_size_t count;
+    uvwasi_size_t size;
+    uvwasi_environ_sizes_get(WASI::g_uvwasi, &count, &size);
+
+    uint32_t* uvEnviron = reinterpret_cast<uint32_t*>(get_memory_pointer(instance, argv[0], count * sizeof(uint32_t)));
+    char* uvEnvironBuf = reinterpret_cast<char*>(get_memory_pointer(instance, argv[1], size));
+
+    if (uvEnviron == nullptr || uvEnvironBuf == nullptr) {
+        result[0] = Value(WasiErrNo::inval);
+        return;
+    }
 
-    char** uvEnviron = reinterpret_cast<char**>(instance->memory(0)->buffer() + env);
-    char* uvEnvironBuf = reinterpret_cast<char*>(instance->memory(0)->buffer() + environBuf);
+    TemporaryData pointers(count * sizeof(void*));
```

Comment:
It seems that `TemporaryData` structure is used only here,
so what about simply allocating a temporal buffer inside this `environ_get` function?

Reply:
#308 should also use this. Sharing code should reduce the possible errors.

Reply:
I see